File: workspace/src/main/java/dev/sayaya/handbook/domain/Group.java
package dev.sayaya.handbook.domain;

import com.fasterxml.jackson.annotation.JsonProperty;
import jsinterop.annotations.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;

@JsType(isNative = true, namespace = JsPackage.GLOBAL, name = "Object")
@Getter(onMethod_ = {@JsOverlay, @JsIgnore})
@Accessors(fluent = true)
@NoArgsConstructor
public final class Group {
    @JsonProperty("id") @JsProperty public String id;
    @JsonProperty("workspace") @JsProperty public String workspace;
    @JsonProperty("name") @JsProperty public String name;
    @JsonProperty("description") @JsProperty public String description;

    @JsOverlay @JsIgnore
    public static Group create(String id, String workspace, String name, String description) {
        if (id == null || id.trim().isEmpty()) throw new IllegalArgumentException("ID cannot be empty");
        if (workspace == null || workspace.trim().isEmpty()) throw new IllegalArgumentException("Workspace ID cannot be empty");
        if (name == null || name.trim().isEmpty()) throw new IllegalArgumentException("Name cannot be empty");
        Group group = new Group();
        group.id = id;
        group.workspace = workspace;
        group.name = name;
        group.description = description;
        return group;
    }
}

File: workspace/src/main/kotlin/dev/sayaya/handbook/domain/Group.kt
package dev.sayaya.handbook.domain

import java.util.*

/**
 * 특정 워크스페이스에 속한 그룹을 나타내는 엔티티(Entity).
 * 그룹의 동등성은 고유 식별자인 `id`를 기준으로 판단합니다.
 */
@JvmRecord
data class Group (
    val id: UUID,
    val workspace: UUID,
    val name: String,
    val description: String?,
) {
    init {
        require(name.isNotBlank()) { "Name cannot be blank" }
    }

    /**
     * 데이터 클래스의 기본 동작(모든 프로퍼티 비교)을 따르지 않고,
     * 엔티티의 동등성 규칙(ID 기반)을 적용하기 위해 재정의합니다.
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Group
        return id == other.id
    }

    /**
     * equals가 id만 비교하므로, hashCode도 id를 기반으로 생성합니다.
     */
    override fun hashCode(): Int {
        return id.hashCode()
    }
}